File: app/lib/core/format.dart
import 'package:intl/intl.dart';

/// Utilitários de formatação (datas em pt_BR, minutos → h/min).
class Fmt {
  static final _diaMes = DateFormat("EEEE, d 'de' MMMM", 'pt_BR');
  static final _curto = DateFormat('dd/MM/yyyy', 'pt_BR');

  static String dataLonga(DateTime d) {
    final s = _diaMes.format(d);
    return s[0].toUpperCase() + s.substring(1);
  }

  static String dataCurta(DateTime d) => _curto.format(d);

  static String minutos(int min) {
    if (min <= 0) return '0 min';
    final h = min ~/ 60;
    final m = min % 60;
    if (h == 0) return '$m min';
    if (m == 0) return '${h}h';
    return '${h}h ${m}min';
  }

  /// yyyy-MM-dd sem depender de locale.
  static String iso(DateTime d) =>
      '${d.year.toString().padLeft(4, '0')}-'
      '${d.month.toString().padLeft(2, '0')}-'
      '${d.day.toString().padLeft(2, '0')}';
}

File: web/src/lib/format.ts
// Utilitários de formatação (espelham lib/core/format.dart).

/** Normaliza pra busca: minúsculas e sem acento ("Português" -> "portugues"). */
export function normalizar(s: string): string {
  return s
    .normalize("NFD")
    .replace(/[̀-ͯ]/g, "")
    .toLowerCase();
}

export function contem(alvo: string, busca: string): boolean {
  if (!busca.trim()) return true;
  return normalizar(alvo).includes(normalizar(busca));
}

/** yyyy-MM-dd sem depender de fuso/locale. */
export function iso(d: Date): string {
  const y = d.getFullYear().toString().padStart(4, "0");
  const m = (d.getMonth() + 1).toString().padStart(2, "0");
  const day = d.getDate().toString().padStart(2, "0");
  return `${y}-${m}-${day}`;
}

/** Parseia "yyyy-MM-dd" como data local (evita off-by-one de fuso do Date ISO). */
export function parseIso(s: string): Date {
  const [y, m, d] = s.split("-").map(Number);
  return new Date(y, (m ?? 1) - 1, d ?? 1);
}

export function addDays(d: Date, n: number): Date {
  const out = new Date(d);
  out.setDate(out.getDate() + n);
  return out;
}

export function diffDays(a: Date, b: Date): number {
  const ms = new Date(a.getFullYear(), a.getMonth(), a.getDate()).getTime() -
    new Date(b.getFullYear(), b.getMonth(), b.getDate()).getTime();
  return Math.round(ms / 86400000);
}

const DIA_MES = new Intl.DateTimeFormat("pt-BR", {
  weekday: "long",
  day: "numeric",
  month: "long",
});

export function dataLonga(d: Date): string {
  const s = DIA_MES.format(d);
  return s.charAt(0).toUpperCase() + s.slice(1);
}

export function dataCurta(d: Date): string {
  return new Intl.DateTimeFormat("pt-BR", {
    day: "2-digit",
    month: "2-digit",
    year: "numeric",
  }).format(d);
}

export function mesAno(d: Date): string {
  const s = new Intl.DateTimeFormat("pt-BR", { month: "long", year: "numeric" }).format(d);
  return s.charAt(0).toUpperCase() + s.slice(1);
}

export function minutos(min: number): string {
  if (min <= 0) return "0 min";
  const h = Math.floor(min / 60);
  const m = min % 60;
  if (h === 0) return `${m} min`;
  if (m === 0) return `${h}h`;
  return `${h}h ${m}min`;
}

export function moeda(v: number): string {
  return new Intl.NumberFormat("pt-BR", { style: "currency", currency: "BRL" }).format(v);
}
